Add `venv` instructions for native installs

mwest1066 opened the pull request.

mwest1066 commented:
This may well be redundant with #10984

eliotwrobson commented:
@mwest1066 It looks like it will be a little while before #10984 will be merged because of some issues that might affect courses. Do you think we should merge this sooner in the meantime?

mwest1066 commented:
Can't hurt to merge this. At least the `.gitignore` change will make my life better.

nwalters512 commented:
In #11033 I picked `.venv/` as the canonical venv path. We could swap that to `venv/`, or support both? @eliotwrobson any idea which one of these is more conventional?

eliotwrobson commented:
I feel like `.venv/` is more conventional. IMO you definitely want a folder that's hidden in the terminal output.

mwest1066 commented on docs/installingNative.md:
````diff
@@ -20,7 +20,7 @@ Most of these prerequisites can be installed using the package manager of your O
 
     ```sh
-    sudo apt install git gcc libc6-dev graphviz graphviz-dev redis6 postgresql15 postgresql15-server postgresql15-contrib
+    sudo apt install git gcc libc6-dev graphviz graphviz-dev redis6 postgresql16 postgresql16-server postgresql16-contrib
````
We are now using postgres 16 in prod. I didn't test this on Ubuntu.

mwest1066 commented on docs/installingNative.md:
````diff
@@ -32,7 +32,7 @@ Most of these prerequisites can be installed using the package manager of your O
 
     ```sh
-    brew install git graphviz postgresql@15 redis@6.2
+    brew install git graphviz postgresql redis uv d2
````
We need postgres >= 16. Installing a specific version with `brew install postgresql@16` means that we need to manually add a PATH to `.zshrc`, which is kind of ugly. I prefer to install the latest version (at the moment that's postgres 17) and just live with the fact that we are running a newer version locally.

We can also use `brew` to install `uv` and `d2`.

reteps replied:
I think the version should be pinned for postgresql, even if the PATHing is a pain.

nwalters512 replied:
I agree with Matt here.

> Mainly because postgres in my understanding (I could be wrong) doesn't make a lot of guarantees of compatibility across versions.

For the data format on disk, sure, but that's hardly relevant here. For syntax/functionality used in queries, they basically always maintain backwards-compatibility. I also run the latest version of Postgres in development and don't experience issues. Of course, we should be careful that we don't use new syntax/features that isn't available in Postgres 16, but (if we happen to do so, and we haven't yet) that would be easily caught in testing or CI.

mwest1066 commented on docs/installingNative.md:
````diff
@@ -71,22 +71,11 @@ Now you can install the other dependencies.
-    ```
-
-    d2 can be installed through the install script:
-
-    ```sh
-    curl -fsSL https://d2lang.com/install.sh | sh -s --
+    brew install node npm
+    corepack enable
````
The [`yarn` docs](https://yarnpkg.com/getting-started/install) prefer installing via `corepack`.

nwalters512 replied:
Node will be removing `corepack` in the future: https://github.com/nodejs/package-maintenance/pull/606. I'm fine to use it until Yarn changes their recommendation.

mwest1066 commented on docs/installingNative.md:
```diff
@@ -130,43 +119,48 @@ Now you can install the other dependencies.
+- Set up a Python virtual environment:
 
-    It is recommended to use a virtual environment for Python dependencies. You can create a virtual environment within PrairieLearn:
-
-    === "Native"
+  === "uv"
```
I made `uv` the first option here. I think it's currently the best way to install the `venv` and should be our recommended package management tool for python.

mwest1066 commented on docs/installingNative.md:
````diff
@@ -130,43 +119,48 @@ Now you can install the other dependencies.
         ```sh
-        python3.10 -m venv venv
-        source venv/bin/activate
+        uv venv --python 3.10 --seed
````
`--seed` is necessary to install `pip` into the new `venv`

reteps replied:
TIL!

mwest1066 commented on docs/installingNative.md:
```diff
@@ -130,43 +119,48 @@ Now you can install the other dependencies.
-        python3.10 -m venv venv
-        source venv/bin/activate
+        uv venv --python 3.10 --seed
+        source .venv/bin/activate
```
Fixing a typo in the existing instructions.

mwest1066 commented on docs/installingNative.md:
````diff
@@ -130,43 +119,48 @@ Now you can install the other dependencies.
 
+  You can run `deactivate` to exit the virtual environment, and `source .venv/bin/activate` to re-enter it.
+
+- On macOS, set the following environment variables so that `pygraphviz` can [find the necessary headers](https://github.com/pygraphviz/pygraphviz/blob/main/INSTALL.txt):
+
+  ```sh
+  cat <<EOF >> .venv/bin/activate
+  export CFLAGS="-I$(brew --prefix graphviz)/include"
+  export LDFLAGS="-L$(brew --prefix graphviz)/lib"
+  EOF
+  source .venv/bin/activate
+  ```
+
````
Appending the `export` commands to the `activate` script like this means that they will automatically apply when we re-run `make deps` anytime in the future, including for package upgrades.

nwalters512 replied:
Huh, interesting solution!

mwest1066 replied:
Suggested to me by ChatGPT 😅

mwest1066 commented on docs/installingNative.md:
````diff
@@ -130,43 +119,48 @@ Now you can install the other dependencies.
+
 - Install all dependencies and transpile local packages:
 
   ```sh
-  # This one command will do everything!
   make deps
+  ```
+
+  The above command installs everything. Alternatively, you can run each step individually:
 
-  # Alternatively, you can run each step individually:
+  ```
````
I separated out `make deps` from the component steps so that you can use the copy button on the code snippet in the docs and get just the one you want (before you would copy both `make deps` and the component commands).

mwest1066 commented:
@nwalters512 @eliotwrobson Thanks for pointing out that we've already switched to `.venv/`. I agree this is better than `venv/` in any case. I also realized that @reteps changed the native install instructions significantly since I wrote this PR, so most of it was now outdated or irrelevant.

I reworked this PR to be updates to @reteps' new version. The main things I changed are:
* Make `uv` the preferred way of creating the virtual environment.
* Prefer using `brew` to install dependencies when possible.
* Don't specify package versions with `brew` but instead use the default/latest version. This has the downside that we'll have a newer version than production (e.g., postgres 17 when prod is 16, node 23 when prod is 20). But if we specify a version with `brew` then we need to manually add the PATH to `.zshrc` which is ugly, and I think it works fine to use newer versions locally (I do this myself in any case).

eliotwrobson reviewed (approved):
Everything seemed reasonable

reteps commented on docs/installingNative.md:
```diff
@@ -71,22 +71,11 @@ Now you can install the other dependencies.
 
```
Should probably update the yarn through npm section on ubuntu as well to mirror `corepack`.

mwest1066 replied:
Done in 321d07d99fa8c01b79a45732c944bff110987f84

reteps commented on docs/installingNative.md:
````diff
@@ -130,43 +119,48 @@ Now you can install the other dependencies.
+        source .venv/bin/activate
         ```
 
+  You can run `deactivate` to exit the virtual environment, and `source .venv/bin/activate` to re-enter it.
````
Unrelated, but I currently use `mise` for activating/deactivating my venv anytime I enter/leave the directory. Maybe a future PR can add instructions for using `direnv` or `mise` to do this.

mwest1066 replied:
Neat! Adding docs for this in a followup PR would be great.

reteps commented on docs/installingNative.md:
```diff
@@ -130,43 +119,48 @@ Now you can install the other dependencies.
+  export CFLAGS="-I$(brew --prefix graphviz)/include"
+  export LDFLAGS="-L$(brew --prefix graphviz)/lib"
+  EOF
+  source .venv/bin/activate
```
nit: either add a newline here or two code block to indicate that this is doing two different operations: 1 should be run only once, but the source should run every time.

mwest1066 replied:
Done in 9dec613a82daba80691e05e443c07e609caa0234

reteps reviewed (approved):
Looks good, left a few notes.

nwalters512 commented on docs/installingNative.md:
```diff
@@ -71,22 +71,11 @@ Now you can install the other dependencies.
 === "macOS"
 
-    Brew can install the rest of the dependencies.
+    Brew can install the rest of the dependencies. Enabling `corepack` makes `yarn` available.
```
It's not clear to me why we separate things into "prerequisites" and "rest of the dependencies", especially since the bulleted list is labeled "prerequisites" but also contains the "other dependencies" (Node). Consider merging them.

mwest1066 replied:
Combined in abbf17d9086180bf6b102f050e23f7e070d66274

nwalters512 commented on docs/installingNative.md:
```diff
@@ -71,22 +71,11 @@ Now you can install the other dependencies.
 
 === "macOS"
```
UI nit for @reteps, though it's not directly related to this PR: for tabbed interfaces like this, the theme doesn't make it easy to tell when the content controlled by the tabs ends (I know there's an `<hr>`, IMO that's not sufficient). Can we improve this? For instance, I like the styles that the [Mise docs](https://mise.jdx.dev/getting-started.html) use:

<img width="717" alt="Screenshot 2025-01-27 at 11 55 09" src="https://github.com/user-attachments/assets/2a93b397-0520-4e7f-9c77-418af05abbac" />

nwalters512 replied:
https://github.com/PrairieLearn/PrairieLearn/pull/11239
